### PR TITLE
Add a renew-cert endpoint

### LINKED ISSUE #58: Clarify "refresh" in certificate response.
Hi there,
In 5.4, the current version of the spec contains a 'refresh'-URI where the client can fetch a new certificate  via plain GET.
I think this is a great feature to prevent the abuse of stolen certificates/private keys, but currently it's not clear how the response should look like.
a) a PEM-encoded certificate could be send
b) a plain DER certificate could be send
c) the whole 'certificate'-Message could be send, containing a new certificate.
d) a HTTP-404/301/whatever could be sent.
e) According to the content-type, different types can be sent.
I definitely prefer c). This makes it easier for CAs (they may change the CA-Chain and/or refresh-URI if they perform changes on their infrastructure) and even the Client development (the 'certificate'-responder may be reused).

Additionally, I'm wondering what will happen if the certificate was revoked in the meanwhile.
To specify this, I'd suggest adding that the server may send the following status codes:
a) 200 (everything's ok)
b) 304 (the certificate was not changed)
c) 302/303 (the cert can be temporarily fetched from there)
d) 307/308 (the cert can be fetched from there & the refresh-uri should be set to the value)
e) 404 (like everywhere else)
g) 403/410 (if the certificate was revoked)
f) 503 (the client should retry later)

### LINKED ISSUE #59: Clarify certificate refresh process
Right now we specify certificate refresh as a GET URL, in response to which the server may choose to provide a refreshed certificate.

It's not specified if the server would issue the refreshed certificate in real-time. I suggest that we should specify that the server will never generate the refreshed certificate in response to /acme/cert/asdf.

 - Because certificate issuance may take more than a few seconds, which means it is longer than we want a GET request to be pending for.
 - Because, as a GET URL, it is not authenticated by the private key. That means that a third party could trigger a certificate issuance even if the account owner did not request it and is not installing the fresh certificate. This is probably not an issue, but could generate strange corner cases.

Instead, the server may choose at its discretion to periodically refresh certain certificates and have them ready to respond to GET /acme/cert/asdf. Response time would be more or less constant and there would be no risk that third party could mutate state with a GET request.

### LINKED ISSUE #125: Renewals and changing cert URLs
@pde said in a different issue:
> Serial numbers aren't and can't be constant. But the spec expects the Certificate URL to be constant across renewals :/.
With a link to the section on Certificate Refresh: https://github.com/letsencrypt/acme-spec/blob/master/draft-barnes-acme.md#certificate-refresh

This is good point. Proposed simple tweak: The cert's GET URL returns the cert, always. If there's an updated version, the response also includes a link relation pointing to the updated cert.

### PR DESCRIPTION
**Pull request by bifurcation**
Certificate renewal should be client-initiated, and should follow the pattern we established with revoke-cert.  (Though we should refer to the certificate by reference, rather than by value.)

This is currently WIP, pending rebase on #138.

Closes #58
Closes #59 
Closes #125

### CODE REVIEW

File: draft-barnes-acme.md [hunk 1]
```diff
@@ -765,18 +767,37 @@ The server provides metadata about the certificate in HTTP headers.  In particul
 HTTP/1.1 201 Created
 Content-Type: application/pkix-cert
 Link: <https://example.com/acme/ca-cert>;rel="up";title="issuer"
-Link: <https://example.com/acme/revoke-cert>;rel="revoke";title="revoke-cert"
+Link: <https://example.com/acme/revoke-cert>;rel="revoke"
```
Comment (jcjones) [inline draft-barnes-acme.md]: Why are we selectively removing the title?

**Comment by jcjones** [comment]
If a server rejects the refresh because the client has asked too recently, should the server give a retry time in the response?

File: draft-barnes-acme.md (hunk 1 above)
Comment (bifurcation) [inline draft-barnes-acme.md]: Because it was causing the line to be too wide for the RFC text format.

**Comment by bifurcation** [comment]
Giving a reason for rejection seems like a good job for a problem document / error code.  Can we handle that in a follow-on?

**Comment by jcjones** [comment]
Sure. Follow-on is fine. I'm good with this one. Marking r=jcjones.

**Comment by martinthomson** [comment]
I think that this is too much machinery.  I think that it is desirable to have renewal as simple as retrieving a resource (GET!).  The simplest model that could work is to replace the contents of the existing certificate resource with updated content at the server's discretion.  But that isn't ideal for other reasons.

If you want to maintain the one issuance -> one resource mapping, which I think is a good design goal, then you have other options:

1. Have a URI that references all certificates, that changes its contents over time. Content-Location header fields can indicate the URI you use for the 1-1 issuance-resource mapping.  We've already talked about having this level of decoupling for the purposes of providing a usable link to clients prior to certificate issuance completing.  Boulder currently issues inline and uses the certificate sequence number to identify the resource, which is an untenable design.  That design could still be used for the 1-1 resource, with an added layer of redirection for this purpose.

2. Use link relations: link to the next certificate from the previous one.  Again, the new certificate would be generated at the discretion of the CA, probably no earlier than half the lifetime of the server.

3. Both 1 and 2.

**Comment by bifurcation** [comment]
Yeah, I don't think we can renewal be done unilaterally by the server.  So we need some request to trigger it.  And I think the URL needs to change; I like the idea of a unique URL per certificate.  With those requirements, I think your proposed alternatives actually end up being more complex from the client POV.  

With the current solution, renewal is a one-liner that's the same every time:
> curl --data "{\"base\":\"${CERT_URL}\"}" $RENEW_URL

With your solution (2), you have to keep track of a different URL to GET every time.  Solution (1) looks comparably simple from a client POV, but it introduces an extra layer of indirection that seems confusing to me.

Unless you see a reason why the proposed approach doesn't work, I'm inclined to proceed.

**Comment by martinthomson** [comment]
Solution (1) requires that the client remember the URL it was given, no more.  And the GET request that it issues to that URL is sufficient to trigger the creation of a new certificate, though CA policy might dictate returning an old one.  This request need never fail once the first certificate is issued.

This is a common design, and - as I pointed out - likely necessary to deal with the problem we discussed previously where the CA needs to give the client a location where it can later retrieve its certificate, but you also want to have an easily enumerable space for issued certificates.  

In other words, the server might want to use sequence numbers to identify certificates once they are issued, but you still need a way to identify the certificate for retrieval before that happens.

**Comment by jsha** [comment]
Can we simplify the protocol by not having a separate renewal / refresh facility? The current letsencrypt client implements renewal using the `new-cert` resource, and it's fine.

The arguments for a special renewal facility, as I understand them:

1. For short-lived certs, renewal may happen once a day and we don't want to have to do authorizations once a day.

However, a client can request a new certificate without reauthorizing, so long as the authorizations are still valid. And a client can check its currently valid authorizations.

2. Renewal without the account key allows the account key to be kept in cold storage.

This is appealing, but unless we allow indefinite renewals (past the expiration of the authorizations), a client will eventually have to do a re-authorization, which means it will require the account key in hot storage. Since we'd like that re-authorization to happen automatically, in the general case this doesn't necessarily accomplish what we want.

I do see some use in the managed hosting case, where one account key may hold authorizations for many domains, but in that case it is probably more appropriate to hold the account key on a separate machine (i.e. not the host that will serve the issued cert), and push validation data to individual hosts as necessary.

**Comment by bifurcation** [comment]
@jsha - I'm not totally sure I agree with your point about hot storage.  If renewal can be done without the account key, then the account key only needs to come out when the authorization expires, which is O(1yr).

Net of the above discussion, I think I like @martinthomson's proposal (1) the best:
* The Location URI you get back effectively identifies the CSR you used
* Actual, specific certificate URIs are provided in Content-Location
* Refresh is GET to certificate URI (from the Location header)
* Response is same as to new-cert 

That meets the "one-liner" criterion, still gets unique URIs per cert, and the indirection isn't *too* confusing.  We might have to think a little bit about how to implement this in boulder, but it seems surmountable.

**Comment by bifurcation** [comment]
With the last change:

Closes #162

File: draft-barnes-acme.md [hunk 2]
```diff
@@ -754,7 +754,17 @@ The values provided in the CSR are only a request, and are not guaranteed.  The
...
+
+~~~~~~~~~~
+
+HTTP/1.1 201 Created
+Location: https://example.com/acme/cert/asdf
+Retry-After: 120
+
+~~~~~~~~~~
+
+If the CA has not yet issued the certificate, the body of this response will be empty, and MUST include a Retry-After header to indicate when the server believes the certificate will be issued (as in the example above).  After that time, the client may send a GET request to the certificate URI in order to get the certificate.
```
Comment (martinthomson) [inline draft-barnes-acme.md]: Retry-After here doesn't work unfortunately.  The header field applies to the request that was made; but clearly this one worked.  You need to put that header field on the subsequent request, which is a shame because you then end up wasting a request with close to 100% probability.

Maybe my understanding isn't entirely correct, so we should ask around about this.  On balance, I think that I'd submit with this in order to solicit feedback.

File: draft-barnes-acme.md [hunk 3]
```diff
@@ -762,21 +772,30 @@ The server provides metadata about the certificate in HTTP headers.  In particul
...
+
+If the CA allows a certificate to be renewed, then it publishes renewed versions of the certificate through the same certificate URI.  Clients retrieve renewed versions of the certificate using a GET query to the certificate URI, which the server should then return in a 200 (OK) response.  The server SHOULD provide a URI for each specific certificate in the Content-Location header field, as shown above.
 
-To refresh the certificate, the client simply sends a GET request to the certificate URL.  This allows the server to provide the client with updated certificates with the same content and different validity intervals, for as long as all of the authorization objects underlying the certificate are valid.
+To avoid unnecessary renewals, the CA may choose not to issue a renewed certificate until it receives such a request.  In such cases, if the CA requires some time to generate the new certificate, the CA MUST return a 202 (Accepted) response, with a Retry-After header field that indicates when the new certifcate will be available.  The CA MAY include the current (non-renewed) certificate as the body of the response.
```
Comment (martinthomson) [inline draft-barnes-acme.md]: 202 + Retry-After is good.

File: draft-barnes-acme.md [hunk 4]
```diff
@@ -762,21 +772,30 @@ The server provides metadata about the certificate in HTTP headers.  In particul
...
 [DER-encoded certificate]
 
 ~~~~~~~~~~
 
-## Certificate Refresh
+## Certificate Renewal
+
+Often, a client wishes to request a new certificate with the same contents as another certificates, but with updated notBefore and notAfter dates.  This operation is referred to as "renewal" of the certificate.
+
+If the CA allows a certificate to be renewed, then it publishes renewed versions of the certificate through the same certificate URI.  Clients retrieve renewed versions of the certificate using a GET query to the certificate URI, which the server should then return in a 200 (OK) response.  The server SHOULD provide a URI for each specific certificate in the Content-Location header field, as shown above.
```
Comment (martinthomson) [inline draft-barnes-acme.md]: ... Requests to a specific certificate resource will always produce the same certificate.

**Comment by martinthomson** [comment]
LGTM, modulo the concerns about the use of Retry-After and 201, which I think need some more consideration.  Though I'd still merge this.

File: draft-barnes-acme.md (hunk 2 above)
Comment (bifurcation) [inline draft-barnes-acme.md]: I was afraid you would say that.  The alternative is to have the client follow up with a GET to the cert URI, resulting in a 202 + Retry-After.  It just seemed like this would optimize out the GET, but I think we actually need it.
